### ISSUE TITLE
Predict forward direction at conditional invertible neural network

### ISSUE THREAD

**Issue by Slewny** [VLL-HD/FrEIA#40]
Hi there!

I trained a **conditional** invertible neural network and am able to predict the inversion of a given label.
At the training step, I used the label as a condition for the network to get the jacobian and latent space z
`z, log_j = net(x, l)`

After that, I optimized the negative log-likelihood loss.

With the given network, I am able to get the inversion of a label by the function:
`net(z, c=l, rev=True)`

with an arbitrary randomized z.

As defined above, the network requires the label as a condition at the training step.
It is possible to predict not only the inversion but also the forward direction without a given label within the conditional network as at a non conditional one?
 
Thanks in advance.

Best regards
Slewny

**Comment by psteinb** [comment]
Good question ... If I understand correctly, in your case the label is used to train the conditioning network. Trying to remove this would imply that the INN is not conditioned any more. 

As the output of the conditioning network is usually concatenated/added to the respective blocks of the INN, you could technically force your conditioning network to output all `0` values and hence not interfere with the INN. However, this is likely to skew the output of the INN as your training was performed *WITH* a non-zero output of the conditioning network and now this has been effectively switched off for inference.

So taking all of the above into account, I would assume that your suggestion is not possible. But @ardizzone or @wapu may know more about this.

**Comment by ardizzone** [comment]
The short answer is sadly "no". But this is not due to software limitations, but fundamentally how the model works.

But if you are mostly interested in sampling (inverse direction), having 'no label', i.e. sampling from p(x) instead of p(x | l), is equivalent to first sampling l_rand from the label distribution, and then inverting the network `net(z, c=l_rand, rev=True)`.

The long answer:
There are two ways to construct a class-conditional likelihood model using INNs.
One of those is to make the INN itself conditional (-> cINN), resulting in p(x | c) := p(z=f(x;c)).
There, it is not possible to run the INN without the label c.
The other possibility, is to keep the standard INN, but make the latent space conditional, i.e. p(x | c) := p(z=f(x) | c).
In the latter case, it is possible to do what you ask. Among other things, this allows the model to be trained in a semi-supervised way, where c is only given some of the time.
Here are some papers that construct their model in the second way:
https://arxiv.org/abs/2001.06448
https://arxiv.org/abs/2001.04872
https://arxiv.org/abs/1406.5298

I hope this helped!

**Comment by Slewny** [comment]
Thank you both for your help! It is understandable and makes it clearer for me, I will take a look into the attached papers then.

Best Regards 
Slewny
